app/ide-desktop/lib/content/src/utils.ts

```diff
@@ -0,0 +1,30 @@
+/** @file Module containing utility functions used throughout this module,
+ * but don't fit anywhere else. */
+
+/** Used to create a "branded type", which is a type intersected with a `Brand<'Name'>`.
+ *
+ * This is useful in parameters that require values from a certain source,
+ * for example IDs for a specific object type.
+ *
+ * It is similar to a `newtype` in other languages,
+ * however because TypeScript is structurally typed, a branded type is assignable to its base type:
+ * `a: string = b as (string & Brand<'Name'>)` successfully typechecks. */
+export interface Newtype<T extends string> {
+	// eslint-disable-next-line @typescript-eslint/naming-convention
+	_$type: T
+}
+
+interface NotNewtype {
+	// eslint-disable-next-line @typescript-eslint/naming-convention
+	_$type?: never
+}
+
+export function asNewtype<T extends Newtype<string>>(s: NotNewtype & Omit<T, '_$type'>): T {
+	// This cast is unsafe. `T` has an extra property `_$type`.
```

Comment:
@wdanilo here is the updated description - not sure if this is any clearer though...

Reply:
1. Ok, so is this unsafe because user couldtheoreticaly use '_$type' in their code? If so (If I understood it correctly now), how is it connected to the sentence that this cast is unsafe? 

2. Also, how does this function work? I understand that it converts any value to new type, right? If so, new type has a field that is a string - but we are not setting this field. I know that this is because of my lack of TS knowledge, but I'm so intrigued by this function now! :D

Reply:
the `_$type` field does not exist at runtime, and is used purely for typechecking.
the reason the unsafe type assertion is needed, is *exactly* because it does not exist at runtime.

maybe i should add that to this comment?

Reply:
Oh it doesn't exist at runtime? I thought that all fields defined by interface need to exist. Yep, adding that to docs would be helpful. Anyway, thanks so much for the explanation!